platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/tipset/ChildlessEventTracker.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.event.tipset;
+
+import com.swirlds.platform.event.EventDescriptor;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Keeps track of events created that have no children. These events are candidates to be used as parents when creating
+ * a new event.
+ */
+public class ChildlessEventTracker {
```

Comment:
This class already exists - `ChatterEventMapper`. Can it be used, maybe with some minor modifications?

Reply:
Keeping this class is fine. But I think the code for keeping only the newest events in the set is better in the other class, because it also supports receiving events out of order.

Reply:
> This data structure is used to track the most recent events from each node.

Ah, now that I think about it, there is a very important difference between these two data structures. `ChatterEventMapper` keeps track of the most recent event from each node, while `ChildlessEventTracker` only cares about keeping track of events that don't have any children yet. This means that the `ChildlessEventTracker` often won't have any events from a particular node.

Suppose you want to build on top of an event `A`. But let's say that somebody beats you to it, and you receive an event `B` that has `A` as an other parent. It's always better to build on top of `B` instead of on top of `A`, since by building on top of `B` you achieve your goal of putting `A` into your ancestry, and get the added benefit of also putting `B` into your ancestry at the same time.

Reply:
updated javadoc